numpy/testing/utils.py

```diff
@@ -1937,3 +1964,150 @@ def __exit__(self, *exc_info):
                 mod.__warningregistry__.clear()
             if mod in self._warnreg_copies:
                 mod.__warningregistry__.update(self._warnreg_copies[mod])
+
+
+class suppressed_warnings(object):
+    """
+    Context manager doing much the same as ``warnings.catch_warnings``.
+    However, it also provides a suppression mechanism, meaning that instead
+    of setting a warning to "ignore", it is possible to add a suppression.
+    This has the advantage of ensuring a clean warning registry.
```

Comment:
Wouldn't be surprised if I have to rewrite half of that doc...

The registry is not cleaned again after leaving the with block, this is only possible if we know the exact module, for any other suppression that does not have `module=blah`, this is impossible to do reliably (or would fun file string -> module stuff). This is to just not change it, no matter what we do inside. (again on new pythons probably not a problem due to invalidation of the registry)